packer/registry/image/image.go

```diff
@@ -0,0 +1,151 @@
+package image
+
+import (
+	"errors"
+	"reflect"
+
+	"github.com/hashicorp/packer-plugin-sdk/packer"
+)
+
+// ArtifactStateURI represents the key used by Packer when querying an packersdk.Artifact
+// for Image metadata that a particular component would like to have stored on the HCP Packer Registry.
+const ArtifactStateURI = "par.artifact.metadata"
+
+// ArtifactOverrideFunc represents a transformation func that can be applied to a
+// non-nil *Image. See WithID, WithRegion functions for examples.
+type ArtifactOverrideFunc func(*Image) error
+
+// Image represents the metadata for some Artifact in the HCP Packer Registry.
+type Image struct {
+	// ImageID is a unique reference identifier stored on the HCP Packer registry
+	// that can be used to get back the built artifact of a builder or post-processor.
+	ImageID string
+	// ProviderName represents the name of the top level cloud or service where the built artifact resides.
+	// For example "aws, azure, docker, gcp, and vsphere".
+	// ProviderRegion represents the location of the built artifact.
+	// For cloud providers region usually maps to a cloud region or zone, but for things like the file builder,
+	// S3 bucket or vsphere cluster region can represent a path on the upstream datastore, or cluster.
+	ProviderName, ProviderRegion string
+	// Labels represents additional details about an image that a builder or post-processor may with to provide for a given build.
+	// Any additional metadata will be made available as build labels within a HCP Packer registry iteration.
+	Labels map[string]string `mapstructure:"labels"`
+}
+
+func New() *Image {
+	return &Image{}
+}
+
+// FromMappedData calls f sequentially for each key and value present in mappedData to create a []*Image
+// as the final return value. If there is an error in calling f, FromMappedData will stop processing mapped items
+// and result in a nil slice, with the said error.
+//
+// FromMappedData will make its best attempt to convert the input map into map[interface{}]interface{} before
+// calling f(k,v). The func f is responsible for type asserting the expected type for the key and value before
+// trying to create an Image from it.
+func FromMappedData(mappedData interface{}, f func(key, value interface{}) (*Image, error)) ([]*Image, error) {
```

Comment:
As an alternative to this implementation which would allow for a map of any type. I'm considering just supporting a mappedData input of map[string]string, as this seems to be the std. map being used across most plugins that return multi-region artifacts. 

It also removes the need for the consumer to do type assertions in the iteration function, which seems a bit complex.

Reply:
Any specific reason you didn't go to `map[string]string` from the beginning?